Feature/flatfile encoder

4hlberg opened the pull request:
Flatfile encoder,

Outstanding: 
* Handling incoming numbers, should it be in the encoder or mapper.
* Finishing readme
* Handling writing of decimals? Possible in flatfiles?

4hlberg committed:
First flatfile write

4hlberg committed:
Add flatfile enc/dec and tests.
Added test files.
Not fixed: data types in write, example if integer needs to be prepended
with zeros

4hlberg committed:
Add Encoder to readme
Started to add examples

4hlberg committed:
add handling of integers on write in encoder.
Makes sure the field is the correct length, also for numbers.

gra-moore commented on README.md:
```diff
@@ -287,3 +288,60 @@ if err != nil {
+        }
+
+    ],
+    "indexFrom":2
```
Is the start position if there is empty space at the beginning of a line? When we discussed this before I thought we landed on just having another field that could have name omitted, and add this as the first column? This would avoid this additional special case.

4hlberg replied:
Could just use the ignore-field already in place and name it, no? That's what I did now, I thought it might be cleaner than adding another check.

gra-moore commented on encoder/flatfile.go:
```diff
@@ -0,0 +1,225 @@
+func NewFlatFileItemWriter(sourceConfig map[string]any, data io.WriteCloser, batchInfo *common_datalayer.BatchInfo) (*FlatFileItemWriter, error) {
+	writer := &FlatFileItemWriter{writer: data, batchInfo: batchInfo}
+	// do this in function or here?
+	itemWriter, err2 := NewFlatFileWriteConfig(sourceConfig["fields"].([]map[string]interface{}), writer)
```
This few lines look a bit off. It would look cleaner if the NewFlatFileWriteConfig returned the config item and then this was assigned into the writer, rather than the side effecting way it is done here. Also, I suspect it works but then the fact that the original writer is returned looks stranges, but also that if there is an error we still returned a writer, when it should probably be nil instead. As another relfection, at the moment there is only one property of the config the Fields property. This makes future extension to add new properties a bit trickier. I suggest having a FlatFileConfig struct with Fields being one property of that. Then instead of fields []FlatFileField on the FlatFileItemWriter there is a config *FlatFileConfig instead. The constructor function for the config takes the sourceConfig.

4hlberg replied:
Fixed, see changes

gra-moore commented on encoder/flatfile.go:
```diff
@@ -0,0 +1,225 @@
+	jsonData, err := json.Marshal(fields)
+	err = json.Unmarshal(jsonData, &config)
+	if err != nil {
+		fmt.Println("Error:", err)
```
Should be replaced with calls to the logger.

4hlberg replied:
returns error

gra-moore commented on encoder/flatfile.go:
```diff
@@ -0,0 +1,225 @@
+	jsonData, err := json.Marshal(fields)
+	err = json.Unmarshal(jsonData, &config)
+	if err != nil {
+		fmt.Println("Error:", err)
```
logger

gra-moore commented on encoder/flatfile.go:
```diff
@@ -0,0 +1,225 @@
+}
+func (c *FlatFileItemWriter) Write(item common_datalayer.Item) error {
+	buf := new(bytes.Buffer)
+	row := item.NativeItem().(map[string]any)
```
while this is ok, it does expose the underlying impl. Suggest if possible to use the exposed functions connected to Item.

gra-moore commented on encoder/flatfile.go:
```diff
@@ -0,0 +1,225 @@
+func NewFlatFileItemIterator(sourceConfig map[string]any, data io.ReadCloser) (*FlatFileItemIterator, error) {
+	scanner := bufio.NewScanner(data)
+	var indexFrom = 0
+	if sourceConfig["indexFrom"] != nil {
```
This maybe isnt needed if you move to have the first column be able to consume these starting bytes.

4hlberg replied:
fixed, see comments above

gra-moore commented on encoder/flatfile.go:
```diff
@@ -0,0 +1,225 @@
+		indexFrom = sourceConfig["indexFrom"].(int)
+	}
+	reader := &FlatFileItemIterator{reader: data, scanner: scanner, indexFrom: indexFrom}
+	itemReader, err2 := NewFlatFileReadConfig(sourceConfig["fields"].([]map[string]interface{}), reader)
```
Again, look at this pattern with regards the comments above.

4hlberg replied:
fixed

gra-moore commented on encoder/flatfile.go:
```diff
@@ -0,0 +1,225 @@
+
+		var step = 0
+		if c.indexFrom != 0 {
+			step = c.indexFrom
```
this can go away if fields are used.

4hlberg replied:
Fixed with ignore

gra-moore commented on encoder/flatfile.go:
```diff
@@ -0,0 +1,225 @@
+				step += field.Length
+				continue
+			} else {
+				entityProps[field.Name] = strings.TrimSpace(line[step : step+field.Length])
```
Would be tempted here to create an item and then use the functions of Item to set the values. Allows, if needed, for impl of Item to change and not affect this part.

4hlberg replied:
leaving for now

gra-moore reviewed (changes requested):
Added some comments

4hlberg committed:
Made updates to config handling
Added flatfile as an option in common datalayer
Made fixes to handling if config is missing

4hlberg committed:
Updated readme with more examples and info
